### PR TITLE
Adding support for Athenz Groups

### PR DESCRIPTION
**Pull request by kpharasi**
@mcieplak @RuiGuo34 @mdhume 

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: pkg/istio/rbac/common/util.go [hunk 1]
```diff
@@ -99,15 +100,57 @@ func (a *ApiHandler) Delete(item *Item) error {
...
+func getMemberExpiry(member interface{}) *rdl.Timestamp {
+	if groupMember, ok := member.(*zms.GroupMember); ok {
+		return groupMember.Expiration
+	}
+
+	if roleMember, ok := member.(*zms.RoleMember); ok {
+		return roleMember.Expiration
+	}
+
+	return &rdl.Timestamp{Time: time.Time{}}
```
[the remark this answers is not in the document]
Comment (RuiGuo34) [reply pkg/istio/rbac/common/util.go:128]: can this simplify to &rdl.Timestamp{}, time.Time{} has a default value of `0001-01-01 00:00:00 +0000 UTC`

[the remark this answers is not in the document]
Comment (RuiGuo34) [reply pkg/istio/rbac/common/util.go:128]: and when will we run into case to return default expiry? Could there be a case when a member is not of type *zms.GroupMember and *zms.RoleMember?

[the remark this answers is not in the document]
Comment (RuiGuo34) [reply pkg/istio/rbac/common/util.go:128]: and when will we run into case to return a default expiry? could there be a case which a member is not of type *zms.GroupMember and *zms.RoleMember

File: pkg/istio/rbac/common/util.go [hunk 2]
```diff
@@ -99,15 +100,57 @@ func (a *ApiHandler) Delete(item *Item) error {
...
+func getMemberSystemDisabled(member interface{}) *int32 {
+	if groupMember, ok := member.(*zms.GroupMember); ok {
+		return groupMember.SystemDisabled
+	}
 
+	if roleMember, ok := member.(*zms.RoleMember); ok {
+		return roleMember.SystemDisabled
+	}
+
+	return nil
```
Comment (RuiGuo34) [inline pkg/istio/rbac/common/util.go:143]: have the same question as above

Comment (kpharasi) [reply pkg/istio/rbac/common/util.go:143]: It should never be the case with our current setup. But this is in case we get something different and to handle that scenario.

File: pkg/istio/rbac/common/util.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kpharasi) [reply pkg/istio/rbac/common/util.go:128]: It should never be the case with our current setup. But this is in case we get something different and to handle that scenario. Will change the default to `nil`

File: pkg/istio/rbac/v2/provider.go [hunk 3]
```diff
@@ -146,8 +146,27 @@ func (p *v2) ConvertAthenzModelIntoIstioRbac(athenzModel athenz.Model, serviceNa
 		from_namespace := &v1beta1.Rule_From{
 			Source: &v1beta1.Source{},
 		}
+
 		// role name should match zms resource name
 		for _, roleMember := range athenzModel.Members[role] {
+			var members []interface{}
+			roleflag := false
```
[the remark this answers is not in the document]
Comment (kpharasi) [reply pkg/istio/rbac/v2/provider.go:153]: This is a flag I have used to differentiate if the type is *zms.GroupMember/*zms.RoleMember and to only run an extra check for expiry and system disabled in it is a *zms.GroupMember (Other than the *zms.RoleMember check)

File: pkg/istio/rbac/common/util.go [hunk 4]
```diff
@@ -246,25 +287,31 @@ func ParseAssertionResource(domainName zms.DomainName, assertion *zms.Assertion)
 
 // CheckAthenzSystemDisabled checks if athenz domain is systematically disabled, if so, controller skips processing current
 // role member
-func CheckAthenzSystemDisabled(roleMember *zms.RoleMember) (bool, error) {
-	if roleMember == nil {
-		return false, fmt.Errorf("got an empty role Member: %s, skipping", roleMember.MemberName)
+func CheckAthenzSystemDisabled(member interface{}) (bool, error) {
+	if member == nil {
+		return false, fmt.Errorf("got an empty role Member: %s, skipping", GetMemberName(member))
```
Comment (mcieplak) [inline pkg/istio/rbac/common/util.go:292]: Should we remove the `%s` from this string and also the GetMemberName call? Since it's nil, it will only return the empty string

File: pkg/istio/rbac/common/util.go [hunk 5]
```diff
@@ -246,25 +287,31 @@ func ParseAssertionResource(domainName zms.DomainName, assertion *zms.Assertion)
...
 }
 
 // CheckAthenzMemberExpiry checks if Expiration field (timezone UTC) is set in the roleMember object, and then
 // checks if expiration date surpasses current time
-func CheckAthenzMemberExpiry(roleMember *zms.RoleMember) (bool, error) {
-	if roleMember == nil {
-		return false, fmt.Errorf("got an empty role Member: %s, skipping", roleMember.MemberName)
+func CheckAthenzMemberExpiry(member interface{}) (bool, error) {
+	if member == nil {
+		return false, fmt.Errorf("got an empty role Member: %s, skipping", GetMemberName(member))
```
Comment (mcieplak) [inline pkg/istio/rbac/common/util.go:307]: Same here as well, won't this just be an empty string?

File: pkg/istio/rbac/common/util_test.go [hunk 6]
```diff
@@ -166,6 +223,101 @@ func TestMemberToOriginJwtSubject(t *testing.T) {
...
+			},
+			expectedName: "client.some-domain.dep-svcA",
+		},
+		{
+			test: "valid group member",
+			member: &zms.GroupMember{
+				MemberName: zms.GroupMemberName("user.somename"),
+			},
+			expectedName: "user.somename",
+		},
```
Comment (mcieplak) [inline pkg/istio/rbac/common/util_test.go:246]: Could we also add a case if member is nil?

File: pkg/istio/rbac/common/util_test.go [hunk 7]
```diff
@@ -166,6 +223,101 @@ func TestMemberToOriginJwtSubject(t *testing.T) {
...
+			member: &zms.GroupMember{
+				MemberName: zms.GroupMemberName("user.somename"),
+				Expiration: &timestamp,
+			},
+			expectedExpiration: &timestamp,
+		},
+	}
+
+	for _, c := range cases {
+		expiry := getMemberExpiry(c.member)
```
Comment (mcieplak) [inline pkg/istio/rbac/common/util_test.go:285]: Can we also add a test case if the member is nil?

File: pkg/istio/rbac/common/util_test.go [hunk 8]
```diff
@@ -166,6 +223,101 @@ func TestMemberToOriginJwtSubject(t *testing.T) {
...
+			member: &zms.GroupMember{
+				MemberName:     zms.GroupMemberName("user.somename"),
+				SystemDisabled: &isNotSystemDisabled,
+			},
+			expectedSystemDisabled: &isNotSystemDisabled,
+		},
+	}
+
+	for _, c := range cases {
+		systemDisabled := getMemberSystemDisabled(c.member)
```
Comment (mcieplak) [inline pkg/istio/rbac/common/util_test.go:316]: Can we also add the nil case here as well?

File: pkg/istio/rbac/v2/provider.go [hunk 9]
```diff
@@ -166,28 +186,57 @@ func (p *v2) ConvertAthenzModelIntoIstioRbac(athenzModel athenz.Model, serviceNa
...
-				log.Errorln("error converting role member to spiffeName: ", err.Error())
-				continue
-			}
-			from_principal.Source.Principals = append(from_principal.Source.Principals, spiffeName)
-			if p.enableOriginJwtSubject {
-				originJwtName, err := common.MemberToOriginJwtSubject(roleMember)
+
+			for _, member := range members {
+				// This is only done for Group Members to check the expiry and system disabled
+				// at a Group Member level after doing a check for the entire role
```
Comment (mcieplak) [inline pkg/istio/rbac/v2/provider.go:192]: Why don't we need to do this for the single role members as well? Little bit confused here as to why we need to differentiate, didn't the previous logic do the check for the role member?

Comment (kpharasi) [reply pkg/istio/rbac/v2/provider.go:192]: So the check outside the loop is for a role member (which is basically a user/service/group). The expiry/system disabled options are either set for this role member or in case of a group it could also be set at the groupmember (all members in the group) level as well. The loop inside detects the scenario where the member is group member and then makes those additional checks. If it was a role member we have already made those checks outside and it would be repeating that. But, in case of groups there are two levels of checks that are required - 1. for the group present as a rolemember in the role and 2. for the particular group in question.

File: pkg/istio/rbac/common/util.go (hunk 5 above)
Comment (kpharasi) [reply pkg/istio/rbac/common/util.go]: True, will change

**Review by mcieplak — APPROVED** [review]

**Review by RuiGuo34 — APPROVED** [review]
